CIP-????/README.md

```diff
@@ -0,0 +1,709 @@
+
+### Costs
+
+The costs of this proposal are substantial.
+
+Plutus Core today has only 8 kinds of term, this proposal adds 2 more, an increase of 25%.
+Additionally, the new term types are the first term types which have a variable number of children.
+Both of these changes increase implementation complexity throughout the system: everything that processes Plutus Core terms must now handle the new cases, and handle the variable numbers of children.
+
+This change also modestly expends our novelty budget.
+Plutus Core is a deliberately conservative language: for the most part it is just the untyped lambda calculus.
+The proposed new features for sums-of-products are also conservative, and follow a very typical pattern.
+But they are not _quite_ as standard as the untyped lambda calculus.
+
+Moreover, a change like this would be very painful to reverse, so we should seriously consider the costs before proceeding.
+
+### Discussion
+
+#### Why is case not strict?
+
+This makes it different to all other Plutus Core terms, but we believe it is the expected behaviour: _no_ language has strict case destructuring.
+
+Consider e.g.
+
+```haskell
+case (Just 1) of 
+   Just x -> print "Success!"
+   Nothing -> error "Failure!"
+```
+
+Even in a strict language like Rust, nobody would expect this to evaluate the `Nothing` branch and evaluate to an error.
+
+Furthermore, this additional laziness has significant performance benefits, see Appendix 2 for more discussion.
+
+#### Is this still worth it if users aren't decoding the script context?
+
+The issues with decoding the script context only apply to users who actually try and decode it; many people have moved on to just using the underlying Data object instead.
+Arguably that saps the motivation for this proposal.
+
+There are two responses to this:
+1. This proposal seems worth it on the basis of its other benefits (benefits 2 and 3).
+2. Removing the problems with decoding the script context may allow users to return to that approach, which can be more ergonomic (and potentially cheaper, since builtin calls are expensive).
+
+#### Why not also make datums and redeemers terms instead of `Data`?
+
+It is always nice to get rid of things, and this CIP does _not_ let us get rid of `Data`, because it is still used for redeemers and datums.
+That begs the question: could we also have redeemers and datums be Plutus Core terms instead of values of type `Data`?
+
+The problem is that, unlike the encoded script context, redeemers and datums are _user facing_.
+They appear in the block format, they can be supplied by users (e.g. directly at the CLI!), and they can be inspected in block explorers.
+There is therefore a significant benefit to having them in a simpler format that is easier to work with.
+If they were Plutus Core terms in generality, we could have all kinds of strange things like people embedding functions in them.
+
+So it still seems beneficial to have redeemers and datums being `Data`.
```

Comment:
I'd challenge this claim @michaelpj, because this removes an extremely interesting property of the current status quo: the homogeneity of a script arguments. At the moment, everything is `Data`, so on-chain script can take one consistent approach to deal with them. This CIP would break that and now demand on-chain script to handle Datum/Redeemer differently from the Script Context. Loosing the consistence here is a massive bummer IMO.

I think we're conflating two things:

(a) the interface in user-facing tools / endpoints
(b) the interface between the ledger and the Plutus VM.

Those do not have to be the same; this coupling is unnecessary / wrong. Said differently, we could maintain `Data` as an interface between the ledger and external software but have the Plutus VM works solely out of Plutus terms. Hence, it is the ledger which would perform this translation from `Data` to a suitable Plutus Term, following the rules outlined in this CIP and how it's already done for the script context. The format in transaction wouldn't change, datums and redeemers will still always be specified as `Data`. Yet, at the moment of evaluating a script, we would translate the `Data` representation to its canonical Plutus term. 

On-chain, a validator would still only ever deal with Plutus Term -- and we could even remove the need for `Data` as a Plutus term (and its associated builtins) **entirely**. 

Without that, we push the handling of this inconsistency down to the on-chain validator; which will irremediably incur an unnecessary overhead.

Reply:
I like this. However it would have a consequence of anything that uses `Data` on-chain.

First thing that comes to mind is the `serialiseData` builtin. I don't have a strong attachment to that particular builtin, but I would be concerned if we cannot serialize and/or hash data on-chain anymore.

Reply:
I'm a little confused about what you're proposing, are you arguing in favour of the alternative discussed in this section, i.e. translate redeemers/datums into SOPs from `Data`?

I do think it's a nice idea in that it lets us get rid of `Data`. I don't think it makes things more consistent unless we _also_ specify the translation of  the script context in two steps, i.e. `scriptContext2Data :: ScriptContext -> Data` and then the generic `data2Term :: Data -> Term`. At the moment we propose to translate the script context directly to a term.

I do find the alternative design of "all the script arguments are `Data`, and they get translated into terms in the same way" to be appealingly consistent. But it does make the argument passing more complex; adds an extra processing step; and possibly locks us into that approach in future. As I said to @L-as , I can't decide if I love it or hate it :sweat_smile: 

> Those do not have to be the same; this coupling is unnecessary / wrong.

They aren't? In fact it seemed to me that the current proposal exactly decouples them: the user-facing stuff (in the transaction) is `Data` and is faithfully represented as such; the ledger-facing stuff uses terms.

Reply:
Another point that I should put into the text: _not_ representing redeemers and datums as `Data` directly makes it harder for script authors, as they now have to know how their data object is encoded in order to actually work with it.

Reply:
> @ch1bo However it would have a consequence of anything that uses Data on-chain.

Not necessarily. Functions like `serializeData` could accept any term instead and return its canonical CBOR representation. The function would be partial (like many other functions in the Plutus world) because not all Plutus terms can be represented as `Data`, but for any legitimate term made of `constr` & builtin types that shouldn't be a problem (except for `String` but that's a special one anyway). 

> @michaelpj I'm a little confused about what you're proposing, are you arguing in favour of the alternative discussed in this section, i.e. translate redeemers/datums into SOPs from Data?

Yes. Precisely. Keep all script arguments consistent. 

> @michaelpj I don't think it makes things more consistent unless we also specify the translation of the script context in two steps. i.e. `scriptContext2Data :: ScriptContext -> Data` and then the generic `data2Term :: Data -> Term`

This is what I am suggesting indeed. Knowing that, we already have the first function since this is what the ledger is currently doing through the Plutus API. So it's only a matter of translating `data2Term` which is -- dare I say -- trivial 😅? Such function can be applied to all arguments given to the CEK machine, so it's only really one function to compose to the existing pipeline. It doesn't change any of the outward facing interfaces. Yet it changes the internal representation. 

> @michaelpj  I can't decide if I love it or hate it.

I can decide for you if you wish <3 

> @michaelpj Another point that I should put into the text: not representing redeemers and datums as Data directly makes it harder for script authors, as they now have to know how their data object is encoded in order to actually work with it.

Well, that's only true if you consider people using Haskell & the PlutusTx framework for off-chain code which abstracts that away behind Template Haskell. Many developers have actually moved to other solutions, might it be Pluto, Plutarch, Helios, Aiken or Plu-ts, etc. And even more are actually using PlutusTx only for on-chain code, and use their own favorite stack for off-chain. In which case, there's a need to bridge the on-chain script interfaces with whatever off-chain solution people use. See for examples projects like Mesh or Lucid that are used for building transactions in TypeScript where it is needed to know how datums and redeemers are encoded in order to build transactions. 

Plus, irrespective of that, tools makers and builders are interested in knowing how the data object is encoded to inspect blockchain traffic and build better interfaces. This is what has motivated [CIP-0057](https://github.com/cardano-foundation/CIPs/pull/258) to begin with. I don't think it's possible to completely hide the encoding format. If this was a initial desire then I can confidently say that it failed given the current situation. This is why I am insisting so much on documenting the interfaces. It is what enables the ecosystem to grow eventually.

Note that I don't think it is a bad situation because it impacts script authors very little. A script author will manipulate types using whatever framework/language they like. For example, they'll deal with Haskell data-types when using PlutusTx and they'll deal with Aiken data-types when using Aiken. Then, it doesn't changes the current status quo where those who use the Haskell Cardano / Plutus APIs down the line will benefits from existing framework functions to build and serialize transactions from Haskell types. While others who already need to manually construct datums and redeemers through their favorite off-chain language will continue doing so. With CIP-0057 we hope to standardize how we document this interface to possibly make this construction of datums and redeemers more automated. CIP-0085 doesn't change that; we'll simply need to extend the specification language from CIP-0057.

Reply:
> So it's only a matter of translating data2Term which is -- dare I say -- trivial

It's not totally trivial because of the `List` and `Map` cases, but it's not too bad.

> Well, that's only true if you consider people using Haskell & the PlutusTx framework for off-chain code which abstracts that away behind Template Haskell.

I think you misunderstood me. I mean that: if you are a script author, you are using some off-chain machinery to generate redeemers and datums as objects of type `Data`. Currently, you manipulate those _as `Data`_ in the script too. Plutus Tx has support to let you use your own data types instead, but fundamentally what's there is the same thing as what you put into the transaction. If we instead convert it into a SOP, then what's there is _not_ the `Data` object you constructed, but rather the result of applying `data2Term` on it. So you now need to know how that works (or rely on some tool to hide that from you). 

> Not necessarily. Functions like serializeData could accept any term instead and return its canonical CBOR representation.

This sounds very complicated and I don't like it. I was barely okay with `seralizeData`.

Reply:
cc @christianschmitz, sorry to pull you in this conversation but since I don't know whether you follow CIPs closely or not I thought you may want to be aware of this. I believe Helios also relies on heavy on the current representation and homogeneity of Plutus `Data` on-chain. 

If you have any new perspective to add to the debate, feel free 🙏